File: Codes in C++/Binary Tree/Check Balanced Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(!root)
            return 0;
        int hl=maxDepth(root->left);
        if(hl==-1)
            return -1;
        int hr=maxDepth(root->right);
        if(hr==-1)
            return -1;
        if(abs(hl-hr)>1)
            return -1;
        return 1+ max(hl,hr);
    }
    bool isBalanced(TreeNode* root) {
        return maxDepth(root)!=-1;
    }
};

File: Codes in Python3/Binary Tree/Check Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def helper(node):
            if not node:
                return 0
            lh=helper(node.left)
            if lh==-1:
                return -1
            rh=helper(node.right)
            if rh==-1:
                return -1            
            if abs(lh-rh)>1:
                return -1
            return 1+max(lh,rh)
        
        return helper(root)!=-1

# TC=O(n)
# SC=O(n)